# Setup

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import  TfidfVectorizer
from scipy.sparse import hstack, vstack
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score, average_precision_score

In [2]:
df = pd.read_csv("csv\\raw_data_labeled_ready.csv")
df = df[df["y"].notnull()]
df.head()

id                                              title  \
0  JyGGMyR3x5I  Machine Learning: Tutorial prático usando apen...   
1  0PrOA2JK6GQ                         Machine Learning Explicado   
2  t5z5lyrb-7s  Introdução ao Machine Learning (ML de Zero a 1...   
3  5q87K1WaoFI  Computer Scientist Explains Machine Learning i...   
4  qDmojpkd8rE  O que é Machine Learning (Aprendizado de Máqui...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 16323674, 'format_...   
1  [{'asr': 48000, 'filesize': 5203175, 'format_i...   
2  [{'asr': 48000, 'filesize': 2603732, 'format_i...   
3  [{'asr': 48000, 'filesize': 9863039, 'format_i...   
4  [{'format_id': '139', 'manifest_url': 'https:/...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  ✅ALURA COM 10% DE DESCONTO: https://www.alura....  2019-11-08   
1  Machine learning (ou aprendizado de máquina) é...  2021-02-25   
2  This video is also available in English: https...  2019-11-13   
3  WIRED has challenged computer scientist and Hi...  2021-08-18   
4  Nesse vídeo vamos aprender o que é machine lea...  2019-06-21   

           uploader               uploader_id  \
0  Filipe Deschamps  UCU5JicSrEM5A63jkJ2QvGYw   
1  Ciência Todo Dia            CienciaTodoDia   
2        TensorFlow  UC0rqucBdTuFTjJiefW5t-IQ   
3             WIRED                     wired   
4     Didática Tech  UC0BiVs5EYh57gzGVvhddjsA   

                                        uploader_url  \
0  http://www.youtube.com/channel/UCU5JicSrEM5A63...   
1         http://www.youtube.com/user/CienciaTodoDia   
2  http://www.youtube.com/channel/UC0rqucBdTuFTjJ...   
3                  http://www.youtube.com/user/wired   
4  http://www.youtube.com/channel/UC0BiVs5EYh57gz...   

                 channel_id  ... tbr  protocol  preference  http_headers  \
0  UCU5JicSrEM5A63jkJ2QvGYw  ... NaN       NaN         NaN           NaN   
1  UCn9Erjy00mpnWeLnRqhsA1g  ... NaN       NaN         NaN           NaN   
2  UC0rqucBdTuFTjJiefW5t-IQ  ... NaN       NaN         NaN           NaN   
3  UCftwRNsjfRo08xYE31tkiyw  ... NaN       NaN         NaN           NaN   
4  UC0BiVs5EYh57gzGVvhddjsA  ... NaN       NaN         NaN           NaN   

   release_date release_year days_since_upload  \
0           NaN          NaN             825.0   
1           NaN          NaN             350.0   
2           NaN          NaN             820.0   
3           NaN          NaN             176.0   
4           NaN          NaN             965.0   

                                    thumbnails_links    y views_per_day  
0  https://i.ytimg.com/vi_webp/JyGGMyR3x5I/maxres...  0.0    219.296970  
1  https://i.ytimg.com/vi/0PrOA2JK6GQ/maxresdefau...  0.0    556.602857  
2  https://i.ytimg.com/vi_webp/t5z5lyrb-7s/maxres...  0.0     17.907317  
3  https://i.ytimg.com/vi/5q87K1WaoFI/maxresdefau...  0.0   5726.215909  
4  https://i.ytimg.com/vi_webp/qDmojpkd8rE/maxres...  0.0     28.239378  

[5 rows x 75 columns]

In [2]:
# features = ["views_per_day", "days_since_upload", "duration", "view_count", "like_count"]
features = ["views_per_day", "days_since_upload", "duration", "view_count"]
# df[features].isnull().sum()

# Train test split

In [4]:
date = "2020-10-05"
Xtrain, Xval = df[df["upload_date"] < date].drop("y", axis=1), df[df["upload_date"] >= date].drop("y", axis=1)
ytrain, yval = df[df["upload_date"] < date]["y"], df[df["upload_date"] >= date]["y"]
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((263, 74), (255, 74), (263,), (255,))

# Title modeling

In [5]:
title_train = Xtrain["title"]
title_val = Xval["title"]

title_vec = TfidfVectorizer(min_df=2)
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)

In [6]:
Xtrain_wtitle = hstack([Xtrain[features], title_bow_train])
Xval_wtitle = hstack([Xval[features], title_bow_val])

In [7]:
Xtrain_wtitle.shape, Xval_wtitle.shape

((263, 225), (255, 225))

In [8]:
mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
mdl.fit(Xtrain_wtitle, ytrain)

RandomForestClassifier(class_weight='balanced', n_estimators=1000, n_jobs=6,
                       random_state=0)

In [9]:
p = mdl.predict_proba(Xval_wtitle)[:, 1]

In [10]:
average_precision_score(yval, p)

0.23206104430364005

In [11]:
roc_auc_score(yval, p)

0.5519785378940308

# active learning

70 samples that the model have more difficult to predict and 30 random choice samples

In [7]:
df_unlabeled = pd.read_csv("csv\\raw_data_labeled_ready.csv")
df_unlabeled = df_unlabeled[df_unlabeled["y"].isnull()].dropna(how="all")
df_unlabeled.shape

ParserError: Error tokenizing data. C error: out of memory

In [13]:
title_u = df_unlabeled["title"]
title_bow_u = title_vec.transform(title_u)

In [14]:
Xu_wtitle = hstack([df_unlabeled[features], title_bow_u])

In [15]:
pu = mdl.predict_proba(Xu_wtitle)[:, 1]

In [16]:
df_unlabeled['p'] = pu

In [17]:
mask_u = (df_unlabeled["p"] >= 0.327) & (df_unlabeled["p"] <= 1.)
mask_u.sum()

71

In [18]:
df_unlabeled[mask_u]

id                                              title  \
463   R0vC31OXt-g  Build, Train and Deploy Machine Learning Model...   
472   lVP7ywRq8p4                   Azure Machine Learning Pipelines   
477   WZ7vS10KPAw  Model deployment and inferencing with Azure Ma...   
479   jGwO_UgTS7I  Lecture 1 - Stanford CS229: Machine Learning -...   
507   x6f5JOPhci0  How optimization for machine learning works, p...   
...           ...                                                ...   
3625  gjOiHCh9CzQ  Jamie Oliver Chooses A Dish For Jamie's Italia...   
3635  8ak3lUxy_yU  How to Make CREAMY MUSHROOM RISOTTO Like an It...   
3648  Nlsm0RlC8zI             2006 WORLD CUP FINAL: Italy 1-1 France   
3652  iLZR7tihHbk  Elena feat. Glance - Mamma mia (He's italiano)...   
3654  7GItOxCYVZQ  ITALIAN ALLSTARS 4 LIFE - MA IL CIELO È SEMPRE...   

                                                formats  \
463   [{'asr': 48000, 'filesize': 13799617, 'format_...   
472   [{'asr': 48000, 'filesize': 4057061, 'format_i...   
477   [{'asr': 48000, 'filesize': 5723810, 'format_i...   
479   [{'asr': 48000, 'filesize': 27046391, 'format_...   
507   [{'asr': 48000, 'filesize': 3892938, 'format_i...   
...                                                 ...   
3625  [{'asr': 48000, 'filesize': 3232706, 'format_i...   
3635  [{'asr': 48000, 'filesize': 6367863, 'format_i...   
3648  [{'asr': 48000, 'filesize': 809985, 'format_id...   
3652  [{'asr': 48000, 'filesize': 1416086, 'format_i...   
3654  [{'asr': 48000, 'filesize': 2194927, 'format_i...   

                                             thumbnails  \
463   [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
472   [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
477   [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
479   [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
507   [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
...                                                 ...   
3625  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3635  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3648  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3652  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3654  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                            description upload_date  \
463   To learn more, please visit: https://aws.amazo...  2018-09-20   
472   This video talks about Azure Machine Learning ...  2019-11-07   
477   In this video, learn about the various deploym...  2021-07-23   
479   For more information about Stanford’s Artifici...  2020-04-17   
507   Part of the End-to-End Machine Learning School...  2018-10-14   
...                                                 ...         ...   
3625  Jamie Oliver visits the MasterChef Canada kitc...  2018-07-25   
3635  My mushroom risotto recipe is a no-fail way to...  2021-05-30   
3648  All the highlights from the famous Final at th...  2014-07-09   
3652  Subscribe to Cat Music Channel: https://goo.gl...  2014-05-21   
3654  ITALIAN ALLSTARS 4 LIFE\n- MA IL CIELO È SEMPR...  2020-05-07   

                     uploader               uploader_id  \
463     AWS Online Tech Talks               AWSwebinars   
472       Microsoft Developer  UCsMica-v34Irf9KVTh6xx-g   
477           Microsoft Azure              windowsazure   
479            stanfordonline            stanfordonline   
507            Brandon Rohrer             BrandonRohrer   
...                       ...                       ...   
3625         MasterChef World  UC_znAZirr11Q5bPrdhKVLZQ   
3635         Vincenzo's Plate            vincenzosplate   
3648                     FIFA                    FIFATV   
3652                Cat Music            catmusicoffice   
3654  ITALIAN ALLSTARS 4 LIFE  UCqe7qlBbYlOufRELKizHQoA   

                                           uploader_url  \
463             http://www.youtube.com/user/AWSwebinars   
472   http://www.youtube.com/channel/UCsMica-v34Irf9

In [19]:
hard = df_unlabeled[mask_u]

In [20]:
random = df_unlabeled[~mask_u].sample(30, random_state=42)

In [21]:
pd.concat([hard, random]).to_csv("csv//active_label.csv", index=False)

In [95]:
!label-studio

^C


# Results of active learning

In [22]:
df_active = pd.read_csv("csv//active_label_labeled.csv")
df_active.head()

Unnamed: 0    id                                              title  \
0         597  3958  AWS Innovate - AI / ML Edition 2022 | How to u...   
1         611  4056  Designing Safe and Reliable Autonomous Systems...   
2        3474  4055  Business Is Booming for the Italian Mafia Duri...   
3        1814  4054  Facebook Stock Plummets 26% in ONE DAY - Lesso...   
4        3020  4053  MASSIVE! Top 4 EXPLOSIVE ASIAN NFT Projects To...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 1903954, 'format_i...   
1  [{'format_id': '139', 'manifest_url': 'https:/...   
2  [{'asr': 48000, 'filesize': 4420653, 'format_i...   
3  [{'asr': 48000, 'filesize': 4154244, 'format_i...   
4  [{'asr': 48000, 'filesize': 5747012, 'format_i...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  Join Telegram [For Free Udemy Courses, Interns...  2022-02-03   
1  Chuchu Fan, Wilson Assistant Professor of Aero...  2022-02-04   
2  The coronavirus hit Italy like a freight train...  2020-08-27   
3  What can we learn from Facebook stocks big mar...  2022-02-03   
4  ✅ FOLLOW ME ON TWITTER 👉https://twitter.com/Do...  2022-02-03   

                  uploader               uploader_id  \
0            Course Intern  UCZaxdoEXp3YvRSO9bcO7S9w   
1  MIT Corporate Relations  UCHWIPmNUwv8PLBpTyOyyP0Q   
2                VICE News                  vicenews   
3    Dave Lee on Investing  UCi2G5aQTN0YwGIfnt1u7Nbg   
4         Donte Walker NFT  UClzzOYzkXuXo2oo1XtsC1JA   

                                        uploader_url  ...  \
0  http://www.youtube.com/channel/UCZaxdoEXp3YvRS...  ...   
1  http://www.youtube.com/channel/UCHWIPmNUwv8PLB...  ...   
2               http://www.youtube.com/user/vicenews  ...   
3  http://www.youtube.com/channel/UCi2G5aQTN0YwGI...  ...   
4  http://www.youtube.com/channel/UClzzOYzkXuXo2o...  ...   

                                    thumbnails_links   y  views_per_day  \
0  https://i.ytimg.com/vi_webp/DnHCPR8GDEE/maxres... NaN      43.428571   
1  https://i.ytimg.com/vi_webp/seBrZhrXeRw/maxres... NaN       2.166667   
2  https://i.ytimg.com/vi_webp/IsNN6d_5oJ8/maxres... NaN    7212.731203   
3  https://i.ytimg.com/vi_webp/U0SCKapxxXc/maxres... NaN   11469.571429   
4  https://i.ytimg.com/vi/_kaU_a4x5Lg/hqdefault.j... NaN     473.857143   

       p              intent  annotator annotation_id  \
0  0.353      É interessante          1           780   
1  0.190  Não é interessante          1           878   
2  0.305  Não é interessante          1           877   
3  0.232  Não é interessante          1           876   
4  0.133  Não é interessante          1           875   

                    created_at                   updated_at  lead_time  
0  2022-02-09T23:56:34.601106Z  2022-02-09T23:56:41.951201Z    114.567  
1  2022-02-10T00:05:12.115902Z  2022-02-10T00:05:12.115902Z      3.139  
2  2022-02-10T00:05:08.466803Z  2022-02-10T00:05:08.466803Z      2.734  
3  2022-02-10T00:05:05.223973Z  2022-02-10T00:05:05.223973Z      2.218  
4  2022-02-10T00:05:02.543396Z  2022-02-10T00:05:02.543396Z      1.498  

[5 rows x 83 columns]

In [23]:
def y(x):
    if x == "É interessante":
        return 1
    if x == "Não é interessante":
        return 0
    return np.nan
df_active["y"] = df_active["intent"].apply(lambda x: y(x))
df_active = df_active.drop(["Unnamed: 0", "created_at", "updated_at", "lead_time", "annotation_id", "annotator", "intent"], axis=1)
df_active.head()

id                                              title  \
0  3958  AWS Innovate - AI / ML Edition 2022 | How to u...   
1  4056  Designing Safe and Reliable Autonomous Systems...   
2  4055  Business Is Booming for the Italian Mafia Duri...   
3  4054  Facebook Stock Plummets 26% in ONE DAY - Lesso...   
4  4053  MASSIVE! Top 4 EXPLOSIVE ASIAN NFT Projects To...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 1903954, 'format_i...   
1  [{'format_id': '139', 'manifest_url': 'https:/...   
2  [{'asr': 48000, 'filesize': 4420653, 'format_i...   
3  [{'asr': 48000, 'filesize': 4154244, 'format_i...   
4  [{'asr': 48000, 'filesize': 5747012, 'format_i...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  Join Telegram [For Free Udemy Courses, Interns...  2022-02-03   
1  Chuchu Fan, Wilson Assistant Professor of Aero...  2022-02-04   
2  The coronavirus hit Italy like a freight train...  2020-08-27   
3  What can we learn from Facebook stocks big mar...  2022-02-03   
4  ✅ FOLLOW ME ON TWITTER 👉https://twitter.com/Do...  2022-02-03   

                  uploader               uploader_id  \
0            Course Intern  UCZaxdoEXp3YvRSO9bcO7S9w   
1  MIT Corporate Relations  UCHWIPmNUwv8PLBpTyOyyP0Q   
2                VICE News                  vicenews   
3    Dave Lee on Investing  UCi2G5aQTN0YwGIfnt1u7Nbg   
4         Donte Walker NFT  UClzzOYzkXuXo2oo1XtsC1JA   

                                        uploader_url  \
0  http://www.youtube.com/channel/UCZaxdoEXp3YvRS...   
1  http://www.youtube.com/channel/UCHWIPmNUwv8PLB...   
2               http://www.youtube.com/user/vicenews   
3  http://www.youtube.com/channel/UCi2G5aQTN0YwGI...   
4  http://www.youtube.com/channel/UClzzOYzkXuXo2o...   

                 channel_id  ... protocol  preference  http_headers  \
0  UCZaxdoEXp3YvRSO9bcO7S9w  ...      NaN         NaN           NaN   
1  UCHWIPmNUwv8PLBpTyOyyP0Q  ...      NaN         NaN           NaN   
2  UCZaT_X_mc0BI-djXOlfhqWQ  ...      NaN         NaN           NaN   
3  UCi2G5aQTN0YwGIfnt1u7Nbg  ...      NaN         NaN           NaN   
4  UClzzOYzkXuXo2oo1XtsC1JA  ...      NaN         NaN           NaN   

   release_date  release_year days_since_upload  \
0           NaN           NaN               7.0   
1           NaN           NaN               6.0   
2           NaN           NaN             532.0   
3           NaN           NaN               7.0   
4           NaN           NaN               7.0   

                                    thumbnails_links  y  views_per_day      p  
0  https://i.ytimg.com/vi_webp/DnHCPR8GDEE/maxres...  1      43.428571  0.353  
1  https://i.ytimg.com/vi_webp/seBrZhrXeRw/maxres...  0       2.166667  0.190  
2  https://i.ytimg.com/vi_webp/IsNN6d_5oJ8/maxres...  0    7212.731203  0.305  
3  https://i.ytimg.com/vi_webp/U0SCKapxxXc/maxres...  0   11469.571429  0.232  
4  https://i.ytimg.com/vi/_kaU_a4x5Lg/hqdefault.j...  0     473.857143  0.133  

[5 rows x 76 columns]

In [24]:
def mapper_y(x):
    y = df_unlabeled.loc[df_unlabeled["thumbnails_links"] == x, "id"].tolist()
    if len(y) != 0:
        return y[0]
    return np.nan

df_active["id"] = df_active["thumbnails_links"].apply(lambda x: mapper_y(x))

In [25]:
average_precision_score(df_active["y"], df_active["p"]), roc_auc_score(df_active["y"], df_active["p"])

(0.262961106290285, 0.46874999999999994)

In [26]:
df_active["new"] = 1

df_total = pd.concat([df, df_active.drop("p", axis=1)])

In [27]:
df_total["new"] = df_total["new"].fillna(0)
df_total.shape

(618, 76)

# Increase validation dataset

In [35]:
date = "2020-10-05"
mask_train = (df_total["upload_date"] < date) & (df_total["new"] == 0)
mask_val = (df_total["upload_date"] >= date) 

Xtrain, Xval = df_total[mask_train].drop("y", axis=1), df_total[mask_val].drop("y", axis=1)
ytrain, yval = df_total[mask_train].pop("y"), df_total[mask_val].pop("y")
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((263, 75), (300, 75), (263,), (300,))

### Title Modeling

In [39]:
title_train = Xtrain["title"]
title_val = Xval["title"]

title_vec = TfidfVectorizer(min_df=2)
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)

Xtrain_wtitle = hstack([Xtrain[features], title_bow_train])
Xval_wtitle = hstack([Xval[features], title_bow_val])

Xtrain_wtitle.shape, Xval_wtitle.shape

((263, 225), (300, 225))

### Function for Validation

In [37]:
def val_score(Xtrain, ytrain, Xval, yval):
    
    mdl = RandomForestClassifier(n_estimators=1000, random_state=0, class_weight="balanced", n_jobs=6)
    mdl.fit(Xtrain, ytrain)
    p = mdl.predict_proba(Xval)[:, 1]
    return average_precision_score(yval, p), roc_auc_score(yval, p) 

In [40]:
val_score(Xtrain_wtitle, ytrain, Xval_wtitle, yval)

(0.224631005125001, 0.5309599999999999)

# Increase train dataset

In [51]:
date = "2020-10-05"
mask_train = (df_total["upload_date"] < date)
mask_val = (df_total["upload_date"] >= date) & (df_total["new"] == 0) 

Xtrain, Xval = df_total[mask_train].drop("y", axis=1), df_total[mask_val].drop("y", axis=1)
ytrain, yval = df_total[mask_train].pop("y"), df_total[mask_val].pop("y")
Xtrain.shape, Xval.shape, ytrain.shape, yval.shape

((318, 75), (255, 75), (318,), (255,))

### Title modeling

In [52]:
title_train = Xtrain["title"]
title_val = Xval["title"]

title_vec = TfidfVectorizer(min_df=2)
title_bow_train = title_vec.fit_transform(title_train)
title_bow_val = title_vec.transform(title_val)

Xtrain_wtitle = hstack([Xtrain[features], title_bow_train])
Xval_wtitle = hstack([Xval[features], title_bow_val])

Xtrain_wtitle.shape, Xval_wtitle.shape

((318, 268), (255, 268))

In [53]:
val_score(Xtrain_wtitle, ytrain, Xval_wtitle, yval)

(0.30226503556733264, 0.5944556226246367)

# Results

                             precision            roc_auc
Without active learning -> 0.2314598372172052, 0.5469483568075116

Increase validation -> 0.224631005125001, 0.5309599999999999

Increase training -> 0.30226503556733264, 0.5944556226246367

In [28]:
df_total.to_csv("csv\\data_ready.csv", index=False)
df_total.head()

id                                              title  \
0  JyGGMyR3x5I  Machine Learning: Tutorial prático usando apen...   
1  0PrOA2JK6GQ                         Machine Learning Explicado   
2  t5z5lyrb-7s  Introdução ao Machine Learning (ML de Zero a 1...   
3  5q87K1WaoFI  Computer Scientist Explains Machine Learning i...   
4  qDmojpkd8rE  O que é Machine Learning (Aprendizado de Máqui...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 16323674, 'format_...   
1  [{'asr': 48000, 'filesize': 5203175, 'format_i...   
2  [{'asr': 48000, 'filesize': 2603732, 'format_i...   
3  [{'asr': 48000, 'filesize': 9863039, 'format_i...   
4  [{'format_id': '139', 'manifest_url': 'https:/...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  ✅ALURA COM 10% DE DESCONTO: https://www.alura....  2019-11-08   
1  Machine learning (ou aprendizado de máquina) é...  2021-02-25   
2  This video is also available in English: https...  2019-11-13   
3  WIRED has challenged computer scientist and Hi...  2021-08-18   
4  Nesse vídeo vamos aprender o que é machine lea...  2019-06-21   

           uploader               uploader_id  \
0  Filipe Deschamps  UCU5JicSrEM5A63jkJ2QvGYw   
1  Ciência Todo Dia            CienciaTodoDia   
2        TensorFlow  UC0rqucBdTuFTjJiefW5t-IQ   
3             WIRED                     wired   
4     Didática Tech  UC0BiVs5EYh57gzGVvhddjsA   

                                        uploader_url  \
0  http://www.youtube.com/channel/UCU5JicSrEM5A63...   
1         http://www.youtube.com/user/CienciaTodoDia   
2  http://www.youtube.com/channel/UC0rqucBdTuFTjJ...   
3                  http://www.youtube.com/user/wired   
4  http://www.youtube.com/channel/UC0BiVs5EYh57gz...   

                 channel_id  ... protocol  preference  http_headers  \
0  UCU5JicSrEM5A63jkJ2QvGYw  ...      NaN         NaN           NaN   
1  UCn9Erjy00mpnWeLnRqhsA1g  ...      NaN         NaN           NaN   
2  UC0rqucBdTuFTjJiefW5t-IQ  ...      NaN         NaN           NaN   
3  UCftwRNsjfRo08xYE31tkiyw  ...      NaN         NaN           NaN   
4  UC0BiVs5EYh57gzGVvhddjsA  ...      NaN         NaN           NaN   

   release_date  release_year days_since_upload  \
0           NaN           NaN             825.0   
1           NaN           NaN             350.0   
2           NaN           NaN             820.0   
3           NaN           NaN             176.0   
4           NaN           NaN             965.0   

                                    thumbnails_links    y views_per_day  new  
0  https://i.ytimg.com/vi_webp/JyGGMyR3x5I/maxres...  0.0    219.296970  0.0  
1  https://i.ytimg.com/vi/0PrOA2JK6GQ/maxresdefau...  0.0    556.602857  0.0  
2  https://i.ytimg.com/vi_webp/t5z5lyrb-7s/maxres...  0.0     17.907317  0.0  
3  https://i.ytimg.com/vi/5q87K1WaoFI/maxresdefau...  0.0   5726.215909  0.0  
4  https://i.ytimg.com/vi_webp/qDmojpkd8rE/maxres...  0.0     28.239378  0.0  

[5 rows x 76 columns]

In [5]:
def mapper_y_new(x):
    y = df_total.loc[df_total["thumbnails_links"] == x, "id"].tolist()
    if len(y) != 0:
        return True
    return False
df_unlabeled["annotated"] = df_unlabeled["thumbnails_links"].apply(lambda x: mapper_y_new(x))

In [6]:
mask = (df_unlabeled["annotated"] == False)

df_unlabeled[mask].sample(400, random_state=42)

ValueError: Cannot take a larger sample than population when 'replace=False'

In [5]:
df_unlabeled = pd.read_csv("csv\\complete_labeled.csv")
df_unlabeled.head()

Unnamed: 0           id                                              title  \
0         230  -G5-kHF57OU  Apa itu Machine Learning? Inilah cara mudah me...   
1         403  6dEUTmoXz0w  Machine Learning With R | Machine Learning Ful...   
2         460  4wneZDEB3VA  Predictive Maintenance & Monitoring using Mach...   
3         461  -g_hyDI4lK4  #03 -What is Machine Learning and Its Applicat...   
4         462  IzcX9bTJLj8          7 Steps to Build a Machine Learning Model   

                                             formats  \
0  [{'format_id': '139', 'manifest_url': 'https:/...   
1  [{'format_id': '139', 'manifest_url': 'https:/...   
2  [{'asr': 48000, 'filesize': 17706117, 'format_...   
3  [{'asr': 48000, 'filesize': 3737826, 'format_i...   
4  [{'asr': 48000, 'filesize': 1401980, 'format_i...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  Machine learning pada dasarnya adalah proses y...  2019-03-17   
1  In this video on Machine Learning with R, you ...  2020-11-14   
2  Learn how to build advanced predictive mainten...  2018-07-24   
3  At Skillari, We believe that Learning is not L...  2022-01-20   
4  Ease data preparation process with organizatio...  2021-04-12   

                           uploader               uploader_id  \
0                     The W Channel                  d3tiuksw   
1                       Simplilearn               Simplilearn   
2                  Google Workspace                GoogleApps   
3  Skillari Machine Learning and AI  UCWlM9hB2TkXuUKIRbH1Y1Ug   
4                       Eye on Tech           TechnologyGuide   

                                        uploader_url  ... http_headers  \
0               http://www.youtube.com/user/d3tiuksw  ...          NaN   
1            http://www.youtube.com/user/Simplilearn  ...          NaN   
2             http://www.youtube.com/user/GoogleApps  ...          NaN   
3  http://www.youtube.com/channel/UCWlM9hB2TkXuUK...  ...          NaN   
4        http://www.youtube.com/user/TechnologyGuide  ...          NaN   

  release_date  release_year  days_since_upload  \
0          NaN           NaN             1061.0   
1          NaN           NaN              453.0   
2          NaN           NaN             1297.0   
3          NaN           NaN               21.0   
4          NaN           NaN              304.0   

                                    thumbnails_links   y views_per_day      p  \
0  https://i.ytimg.com/vi/-G5-kHF57OU/hqdefault.j... NaN     21.480679  0.094   
1  https://i.ytimg.com/vi_webp/6dEUTmoXz0w/maxres... NaN     74.997792  0.177   
2  https://i.ytimg.com/vi_webp/4wneZDEB3VA/maxres... NaN     32.144950  0.110   
3  https://i.ytimg.com/vi_webp/-g_hyDI4lK4/maxres... NaN    676.285714  0.138   
4  https://i.ytimg.com/vi/IzcX9bTJLj8/maxresdefau... NaN      9.263158  0.117   

  annotated          id_  
0     False  -G5-kHF57OU  
1     False  6dEUTmoXz0w  
2     False  4wneZDEB3VA  
3     False  -g_hyDI4lK4  
4     False  IzcX9bTJLj8  

[5 rows x 79 columns]

In [7]:
df_unlabeled["id_"] = df_unlabeled["id"]
df_unlabeled[['id_', 'title', 'thumbnails_links', 'view_count', 'uploader', 'description']].to_csv("csv\\complete_labeled_to_label.csv", index=False)

In [2]:
df2 = pd.read_csv("csv\\complete_labeled_to_label.csv")
df2.head()

id_                                              title    y  \
0       #NOME?  Apa itu Machine Learning? Inilah cara mudah me...  0.0   
1  6dEUTmoXz0w  Machine Learning With R | Machine Learning Ful...  0.0   
2  4wneZDEB3VA  Predictive Maintenance & Monitoring using Mach...  1.0   
3       #NOME?  #03 -What is Machine Learning and Its Applicat...  0.0   
4  IzcX9bTJLj8          7 Steps to Build a Machine Learning Model  1.0   

                                    thumbnails_links  view_count  \
0  https://i.ytimg.com/vi/-G5-kHF57OU/hqdefault.j...       22791   
1  https://i.ytimg.com/vi_webp/6dEUTmoXz0w/maxres...       33974   
2  https://i.ytimg.com/vi_webp/4wneZDEB3VA/maxres...       41692   
3  https://i.ytimg.com/vi_webp/-g_hyDI4lK4/maxres...       14202   
4  https://i.ytimg.com/vi/IzcX9bTJLj8/maxresdefau...        2816   

                           uploader  \
0                     The W Channel   
1                       Simplilearn   
2                  Google Workspace   
3  Skillari Machine Learning and AI   
4                       Eye on Tech   

                                         description  
0  Machine learning pada dasarnya adalah proses y...  
1  In this video on Machine Learning with R, you ...  
2  Learn how to build advanced predictive mainten...  
3  At Skillari, We believe that Learning is not L...  
4  Ease data preparation process with organizatio...

In [3]:
df2[df2.y.notnull()].shape

(501, 7)

In [4]:
df = pd.read_csv("csv\\raw_data_labeled_ready.csv")
df.head()

id                                              title  \
0  JyGGMyR3x5I  Machine Learning: Tutorial prático usando apen...   
1  0PrOA2JK6GQ                         Machine Learning Explicado   
2  t5z5lyrb-7s  Introdução ao Machine Learning (ML de Zero a 1...   
3  5q87K1WaoFI  Computer Scientist Explains Machine Learning i...   
4  qDmojpkd8rE  O que é Machine Learning (Aprendizado de Máqui...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 16323674, 'format_...   
1  [{'asr': 48000, 'filesize': 5203175, 'format_i...   
2  [{'asr': 48000, 'filesize': 2603732, 'format_i...   
3  [{'asr': 48000, 'filesize': 9863039, 'format_i...   
4  [{'format_id': '139', 'manifest_url': 'https:/...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  ✅ALURA COM 10% DE DESCONTO: https://www.alura....  2019-11-08   
1  Machine learning (ou aprendizado de máquina) é...  2021-02-25   
2  This video is also available in English: https...  2019-11-13   
3  WIRED has challenged computer scientist and Hi...  2021-08-18   
4  Nesse vídeo vamos aprender o que é machine lea...  2019-06-21   

           uploader               uploader_id  \
0  Filipe Deschamps  UCU5JicSrEM5A63jkJ2QvGYw   
1  Ciência Todo Dia            CienciaTodoDia   
2        TensorFlow  UC0rqucBdTuFTjJiefW5t-IQ   
3             WIRED                     wired   
4     Didática Tech  UC0BiVs5EYh57gzGVvhddjsA   

                                        uploader_url  \
0  http://www.youtube.com/channel/UCU5JicSrEM5A63...   
1         http://www.youtube.com/user/CienciaTodoDia   
2  http://www.youtube.com/channel/UC0rqucBdTuFTjJ...   
3                  http://www.youtube.com/user/wired   
4  http://www.youtube.com/channel/UC0BiVs5EYh57gz...   

                 channel_id  ... tbr  protocol  preference  http_headers  \
0  UCU5JicSrEM5A63jkJ2QvGYw  ... NaN       NaN         NaN           NaN   
1  UCn9Erjy00mpnWeLnRqhsA1g  ... NaN       NaN         NaN           NaN   
2  UC0rqucBdTuFTjJiefW5t-IQ  ... NaN       NaN         NaN           NaN   
3  UCftwRNsjfRo08xYE31tkiyw  ... NaN       NaN         NaN           NaN   
4  UC0BiVs5EYh57gzGVvhddjsA  ... NaN       NaN         NaN           NaN   

   release_date release_year days_since_upload  \
0           NaN          NaN             825.0   
1           NaN          NaN             350.0   
2           NaN          NaN             820.0   
3           NaN          NaN             176.0   
4           NaN          NaN             965.0   

                                    thumbnails_links    y views_per_day  
0  https://i.ytimg.com/vi_webp/JyGGMyR3x5I/maxres...  0.0    219.296970  
1  https://i.ytimg.com/vi/0PrOA2JK6GQ/maxresdefau...  0.0    556.602857  
2  https://i.ytimg.com/vi_webp/t5z5lyrb-7s/maxres...  0.0     17.907317  
3  https://i.ytimg.com/vi/5q87K1WaoFI/maxresdefau...  0.0   5726.215909  
4  https://i.ytimg.com/vi_webp/qDmojpkd8rE/maxres...  0.0     28.239378  

[5 rows x 75 columns]

In [5]:
def mapper_y(x):
    y = df2.loc[df2["thumbnails_links"] == x, "y"].tolist()
    if len(y) != 0:
        return y[0]
    return df.loc[df["thumbnails_links"] == x, "y"].tolist()[0]

df["y"] = df["thumbnails_links"].apply(lambda x: mapper_y(x))

In [6]:
df_total = pd.read_csv("csv\\data_ready.csv")
df_total.head()

id                                              title  \
0  JyGGMyR3x5I  Machine Learning: Tutorial prático usando apen...   
1  0PrOA2JK6GQ                         Machine Learning Explicado   
2  t5z5lyrb-7s  Introdução ao Machine Learning (ML de Zero a 1...   
3  5q87K1WaoFI  Computer Scientist Explains Machine Learning i...   
4  qDmojpkd8rE  O que é Machine Learning (Aprendizado de Máqui...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 16323674, 'format_...   
1  [{'asr': 48000, 'filesize': 5203175, 'format_i...   
2  [{'asr': 48000, 'filesize': 2603732, 'format_i...   
3  [{'asr': 48000, 'filesize': 9863039, 'format_i...   
4  [{'format_id': '139', 'manifest_url': 'https:/...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  ✅ALURA COM 10% DE DESCONTO: https://www.alura....  2019-11-08   
1  Machine learning (ou aprendizado de máquina) é...  2021-02-25   
2  This video is also available in English: https...  2019-11-13   
3  WIRED has challenged computer scientist and Hi...  2021-08-18   
4  Nesse vídeo vamos aprender o que é machine lea...  2019-06-21   

           uploader               uploader_id  \
0  Filipe Deschamps  UCU5JicSrEM5A63jkJ2QvGYw   
1  Ciência Todo Dia            CienciaTodoDia   
2        TensorFlow  UC0rqucBdTuFTjJiefW5t-IQ   
3             WIRED                     wired   
4     Didática Tech  UC0BiVs5EYh57gzGVvhddjsA   

                                        uploader_url  \
0  http://www.youtube.com/channel/UCU5JicSrEM5A63...   
1         http://www.youtube.com/user/CienciaTodoDia   
2  http://www.youtube.com/channel/UC0rqucBdTuFTjJ...   
3                  http://www.youtube.com/user/wired   
4  http://www.youtube.com/channel/UC0BiVs5EYh57gz...   

                 channel_id  ... protocol  preference  http_headers  \
0  UCU5JicSrEM5A63jkJ2QvGYw  ...      NaN         NaN           NaN   
1  UCn9Erjy00mpnWeLnRqhsA1g  ...      NaN         NaN           NaN   
2  UC0rqucBdTuFTjJiefW5t-IQ  ...      NaN         NaN           NaN   
3  UCftwRNsjfRo08xYE31tkiyw  ...      NaN         NaN           NaN   
4  UC0BiVs5EYh57gzGVvhddjsA  ...      NaN         NaN           NaN   

   release_date  release_year days_since_upload  \
0           NaN           NaN             825.0   
1           NaN           NaN             350.0   
2           NaN           NaN             820.0   
3           NaN           NaN             176.0   
4           NaN           NaN             965.0   

                                    thumbnails_links    y views_per_day  new  
0  https://i.ytimg.com/vi_webp/JyGGMyR3x5I/maxres...  0.0    219.296970  0.0  
1  https://i.ytimg.com/vi/0PrOA2JK6GQ/maxresdefau...  0.0    556.602857  0.0  
2  https://i.ytimg.com/vi_webp/t5z5lyrb-7s/maxres...  0.0     17.907317  0.0  
3  https://i.ytimg.com/vi/5q87K1WaoFI/maxresdefau...  0.0   5726.215909  0.0  
4  https://i.ytimg.com/vi_webp/qDmojpkd8rE/maxres...  0.0     28.239378  0.0  

[5 rows x 76 columns]

In [7]:
def mapper_y(x):
    y = df_total.loc[df_total["thumbnails_links"] == x, "y"].tolist()
    if len(y) != 0:
        return y[0]
    return df.loc[df["thumbnails_links"] == x, "y"].tolist()[0]

df["y"] = df["thumbnails_links"].apply(lambda x: mapper_y(x))

In [8]:
df.head()

id                                              title  \
0  JyGGMyR3x5I  Machine Learning: Tutorial prático usando apen...   
1  0PrOA2JK6GQ                         Machine Learning Explicado   
2  t5z5lyrb-7s  Introdução ao Machine Learning (ML de Zero a 1...   
3  5q87K1WaoFI  Computer Scientist Explains Machine Learning i...   
4  qDmojpkd8rE  O que é Machine Learning (Aprendizado de Máqui...   

                                             formats  \
0  [{'asr': 48000, 'filesize': 16323674, 'format_...   
1  [{'asr': 48000, 'filesize': 5203175, 'format_i...   
2  [{'asr': 48000, 'filesize': 2603732, 'format_i...   
3  [{'asr': 48000, 'filesize': 9863039, 'format_i...   
4  [{'format_id': '139', 'manifest_url': 'https:/...   

                                          thumbnails  \
0  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
1  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
2  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
3  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   
4  [{'height': 94, 'url': 'https://i.ytimg.com/vi...   

                                         description upload_date  \
0  ✅ALURA COM 10% DE DESCONTO: https://www.alura....  2019-11-08   
1  Machine learning (ou aprendizado de máquina) é...  2021-02-25   
2  This video is also available in English: https...  2019-11-13   
3  WIRED has challenged computer scientist and Hi...  2021-08-18   
4  Nesse vídeo vamos aprender o que é machine lea...  2019-06-21   

           uploader               uploader_id  \
0  Filipe Deschamps  UCU5JicSrEM5A63jkJ2QvGYw   
1  Ciência Todo Dia            CienciaTodoDia   
2        TensorFlow  UC0rqucBdTuFTjJiefW5t-IQ   
3             WIRED                     wired   
4     Didática Tech  UC0BiVs5EYh57gzGVvhddjsA   

                                        uploader_url  \
0  http://www.youtube.com/channel/UCU5JicSrEM5A63...   
1         http://www.youtube.com/user/CienciaTodoDia   
2  http://www.youtube.com/channel/UC0rqucBdTuFTjJ...   
3                  http://www.youtube.com/user/wired   
4  http://www.youtube.com/channel/UC0BiVs5EYh57gz...   

                 channel_id  ... tbr  protocol  preference  http_headers  \
0  UCU5JicSrEM5A63jkJ2QvGYw  ... NaN       NaN         NaN           NaN   
1  UCn9Erjy00mpnWeLnRqhsA1g  ... NaN       NaN         NaN           NaN   
2  UC0rqucBdTuFTjJiefW5t-IQ  ... NaN       NaN         NaN           NaN   
3  UCftwRNsjfRo08xYE31tkiyw  ... NaN       NaN         NaN           NaN   
4  UC0BiVs5EYh57gzGVvhddjsA  ... NaN       NaN         NaN           NaN   

   release_date release_year days_since_upload  \
0           NaN          NaN             825.0   
1           NaN          NaN             350.0   
2           NaN          NaN             820.0   
3           NaN          NaN             176.0   
4           NaN          NaN             965.0   

                                    thumbnails_links    y views_per_day  
0  https://i.ytimg.com/vi_webp/JyGGMyR3x5I/maxres...  0.0    219.296970  
1  https://i.ytimg.com/vi/0PrOA2JK6GQ/maxresdefau...  0.0    556.602857  
2  https://i.ytimg.com/vi_webp/t5z5lyrb-7s/maxres...  0.0     17.907317  
3  https://i.ytimg.com/vi/5q87K1WaoFI/maxresdefau...  0.0   5726.215909  
4  https://i.ytimg.com/vi_webp/qDmojpkd8rE/maxres...  0.0     28.239378  

[5 rows x 75 columns]

In [9]:
df.to_csv("csv\\total_data.csv", index=False)